database/database.go

```diff
@@ -153,6 +154,49 @@ func (db DatastoreBacked) StateForDomain(domain string) (state DomainState, err
 	return state, nil
 }
 
+// StatesForDomains returns the domains states for the given domains
+func (db DatastoreBacked) StatesForDomains(domains []string) (states []DomainState, err error) {
+	// Set up datastore context
+	c, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	client, datastoreErr := db.backend.NewClient(c, db.projectID)
+	if datastoreErr != nil {
+		return nil, datastoreErr
+	}
+
+	getDomainStates := func(keys []*datastore.Key) ([]DomainState, error) {
+		domainStates := make([]DomainState, len(keys))
+		getErr := client.GetMulti(c, keys, domainStates)
+		if getErr != nil {
+			return nil, getErr
+		}
+		for i := range domainStates {
+			domainStates[i].Name = keys[i].Name
+		}
+		return domainStates, nil
+	}
+
+	var keys []*datastore.Key
+	var res []DomainState
+	var getErr error
+	for _, domain := range domains {
+		key := datastore.NameKey(domainStateKind, domain, nil)
+		keys = append(keys, key)
+		if len(keys) >= batchSize {
+			tempStates, getErr := getDomainStates(keys)
+			res = append(res, tempStates...)
+			if getErr != nil {
```

Comment:
Nit: Prefer doing the error check immediately after the call to getDomainStates (since the append call doesn't matter if we get an error)